### ISSUE TITLE
StringIndexOutOfBoundsException when calling Context.evaluateString for preloading Java Objects

### ISSUE THREAD

**Issue by heywiki** [mozilla/rhino#270]
We have the following issue using rhino 1.7.7 (also tried 1.7.7.1) when preloading java classes with

```
String loadMe = "RegExp;getClass;java;Packages;JavaAdapter;";
cx.evaluateString(sealedSharedScope , loadMe, "lazyLoad", 0, null);
```

We get the following error 

`java.lang.StringIndexOutOfBoundsException: String index out of range: 3`

This occured after updating the machine from Java 1.8.0_72 to 74, (Java Version is 1.8.0_74-b02, Windows 10, Running inside Tomcat 8.0.32)

However this seems not to be the issue because this member of our team also can reproduce this error in his VMs with different setups. This only happens on his machine. This only happens on his machine.
## Full Stacktrace is

java.lang.StringIndexOutOfBoundsException: String index out of range: 3
    at java.lang.String.charAt(Unknown Source)
    at org.apache.catalina.loader.WebappClassLoaderBase.filter(WebappClassLoaderBase.java:2780)
    at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1253)
    at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1142)
    at org.mozilla.javascript.Kit.classOrNull(Kit.java:60)
    at org.mozilla.javascript.NativeJavaPackage.getPkgProperty(NativeJavaPackage.java:129)
    at org.mozilla.javascript.NativeJavaPackage.get(NativeJavaPackage.java:84)
    at org.mozilla.javascript.NativeJavaTopPackage.init(NativeJavaTopPackage.java:96)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
    at java.lang.reflect.Method.invoke(Unknown Source)
    at org.mozilla.javascript.ScriptableObject.buildClassCtor(ScriptableObject.java:1309)
    at org.mozilla.javascript.LazilyLoadedCtor.buildValue0(LazilyLoadedCtor.java:105)
    at org.mozilla.javascript.LazilyLoadedCtor.access$000(LazilyLoadedCtor.java:18)
    at org.mozilla.javascript.LazilyLoadedCtor$1.run(LazilyLoadedCtor.java:90)
    at java.security.AccessController.doPrivileged(Native Method)
    at org.mozilla.javascript.LazilyLoadedCtor.buildValue(LazilyLoadedCtor.java:86)
    at org.mozilla.javascript.LazilyLoadedCtor.init(LazilyLoadedCtor.java:66)
    at org.mozilla.javascript.ScriptableObject$GetterSlot.getValue(ScriptableObject.java:320)
    at org.mozilla.javascript.ScriptableObject.get(ScriptableObject.java:472)
    at org.mozilla.javascript.IdScriptableObject.get(IdScriptableObject.java:334)
    at org.mozilla.javascript.ScriptableObject.getProperty(ScriptableObject.java:2269)
    at org.mozilla.javascript.ScriptRuntime.topScopeName(ScriptRuntime.java:1947)
    at org.mozilla.javascript.ScriptRuntime.name(ScriptRuntime.java:1849)
    at org.mozilla.javascript.Interpreter.interpretLoop(Interpreter.java:1558)
    at org.mozilla.javascript.Interpreter.interpret(Interpreter.java:815)
    at org.mozilla.javascript.InterpretedFunction.call(InterpretedFunction.java:109)
    at org.mozilla.javascript.ContextFactory.doTopCall(ContextFactory.java:393)
    at org.wikiwizard.juwi.script.JuwiContextFactory.doTopCall(JuwiContextFactory.java:95)
    at org.mozilla.javascript.ScriptRuntime.doTopCall(ScriptRuntime.java:3282)
    at org.mozilla.javascript.InterpretedFunction.exec(InterpretedFunction.java:120)
    at org.mozilla.javascript.Context.evaluateString(Context.java:1219)
    at org.wikiwizard.juwi.script.RhinoEngineWrapper.run(RhinoEngineWrapper.java:162)

**Comment by heywiki** [comment]
We put try catch around it, but now of course the usage of direct packages wont work anymore (throws same exception!)

```
var putil = new Packages.org.wikiwizard.commons.env.PlatformUtil(); 
jout.println(putil.getOSName());
```

Throws java.lang.StringIndexOutOfBoundsException: String index out of range: 3

**Comment by gbrail** [comment]
It's hard to tell what the problem is when the error is coming from another project's class loader. (org.apache.catalina.loader.WebappClassLoaderBase.filter(WebappClassLoaderBase.java:2780) Do you have any way to figure out what argument Rhino is passing to that class, or to reproduce this issue in Rhino without that? Or is this an open-source project that we can somehow test ourselves?

Thanks -- hard to guess at the cause without some more info.

**Comment by pberlowski** [comment]
Hi, I'm experiencing the same issue but related to an update from Tomcat 8.0.26 to 8.0.32. My Java version is unchanged at 1.8.0_74. The stack trace on Tomcat 8.0.32 is identical, but the 8.0.26 works without a complaint.

**Comment by some-ejit** [comment]
I am having the same issue on Tomcat 8.0.32 (embedded) with Java 1.8.0_51 and Rhino 1.7R2. I moved back to Tomcat 8.0.30 and this resolved the issue.

**Comment by getgot** [comment]
I am having the same problem too on Tomcat 8.0.77. I have found the patch with javaassist in http://takahashikzn.root42.jp/entry/2016/02/16/141635 (japanese). but, I don't wont to use javaassist. Does someone have other solution ? 

According to the upper url, it is caused from change of classloader behavior in Tomcat. Ver 8.0.32 and over, Tomcat will throw StringIndexOutOfBoundsException in case that class load fail in some way. Rhino don't catch it in Kit#classOrNull. so we get the exeption.

**Comment by getgot** [comment]
Maybe, This is fixed on side of tomcat8 / 8.0.33, https://bz.apache.org/bugzilla/show_bug.cgi?id=58999.

**Comment by p-bakker** [comment]
Closing this as it doesn't seem related to the Rhino codebase and a fix has apparently been appied to Tomcat for this, according to the previous comment
